### PR TITLE
[PyTorch] Drop FA as an installation requirement

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR removes [flash-attention](https://github.com/Dao-AILab/flash-attention) as a required dependency for Transformer Engine. Users can choose to install FA2 via:
```
# supported versions: >= 2.1.1, <=2.6.3
pip install flash-attn==2.6.3
```
or FA3 via:
```
# supported versions: >= 3.0.0b1
1) pip install "git+https://github.com/Dao-AILab/flash-attention.git#egg=flashattn-hopper&subdirectory=hopper"
2) python_path=`python -c "import site; print(site.getsitepackages()[0])"`
3) mkdir -p $python_path/flashattn_hopper
4) wget -P $python_path/flashattn_hopper https://raw.githubusercontent.com/Dao-AILab/flash-attention/main/hopper/flash_attn_interface.py
```

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Remove flash-attention as a build dependency
- Change logic around backend selection/recommendation in attention

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -86,63 +86,110 @@
...
+if not fa_logger.hasHandlers():
+    fa_logger.addHandler(_stream_handler)
+_flash_attn_installation_steps = """\
+pip install "git+https://github.com/Dao-AILab/flash-attention.git" """
+try:
+    _flash_attn_version = PkgVersion(get_pkg_version("flash-attn"))
+except PackageNotFoundError:
+    if get_device_compute_capability() >= (8, 0) and _NVTE_FLASH_ATTN:
+        fa_logger.debug(
+            "To use flash-attn v2, please use the following commands to install: \n%s",
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:126]:
```suggestion
            "Flash Attention not installed. To use flash-attn v2, please use the following commands to install: \n%s",
```

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -802,6 +867,17 @@ def get_attention_backend(
 
     # All available backends
     available_backends = [use_flash_attention, use_fused_attention, use_unfused_attention]
+    # if not use_fused_attention and use_flash_attention and not _flash_attn_is_installed:
+    if use_flash_attention and not _flash_attn_is_installed:
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:871]: This condition does not do the same thing as the comment says as it does not check if use_fused_attention is false.

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -828,20 +904,20 @@ def get_attention_backend(
...
+    #        use_flash_attention
+    #        and use_fused_attention
+    #        and fused_attention_backend == FusedAttnBackend["FP8"]
+    #        and _use_flash_attn_3
+    #    ):
+    #        logger.debug(
+    #            "Disabling FlashAttention 3 to give FusedAttention preference as FusedAttention "
+    #            "supports more accurate scaling factors in FP8 execution"
+    #        )
+    #        use_flash_attention = False
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:920]: Please remember to clean it up.

Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py]: I've updated the logic and cleaned up a bit. Sorry what you saw was an unfinished version.

File: transformer_engine/pytorch/attention.py [hunk 4]
```diff
@@ -95,39 +96,97 @@
...
 _NVTE_FLASH_ATTN = int(os.getenv("NVTE_FLASH_ATTN", "1"))
 _NVTE_FUSED_ATTN = int(os.getenv("NVTE_FUSED_ATTN", "1"))
 _NVTE_UNFUSED_ATTN = int(os.getenv("NVTE_UNFUSED_ATTN", "1"))
-_flash_attn_version = PkgVersion(get_pkg_version("flash-attn"))
+
+# Detect flash-attn v2 in the environment
+_flash_attn_is_installed = False
+_flash_attn_version = PkgVersion("0")
 _flash_attn_version_required = PkgVersion("2.0.6")
+_flash_attn_version_not_supported = (PkgVersion("2.0.9"), PkgVersion("2.1.0"))
```
Comment (ksivaman) [inline transformer_engine/pytorch/attention.py:125]: I think this is leading to a lot of additional maintenance. Now that we're making FA optional I suggest we move the minimum FA to 2.1.1 and get rid of this logic unless we're committed to supporting 2.0.6 - 2.0.8 (which I don't think is necessary) .

File: transformer_engine/pytorch/attention.py [hunk 5]
```diff
@@ -95,39 +96,97 @@
...
 _stream_handler = logging.StreamHandler()
 _stream_handler.setFormatter(_formatter)
+fa_logger = logging.getLogger()
+fa_logger.setLevel(_log_level)
+if not fa_logger.hasHandlers():
+    fa_logger.addHandler(_stream_handler)
+
+
+@functools.lru_cache(maxsize=None)
+def _get_supported_versions(version_min, version_not_supported, version_max):
```
Comment (ksivaman) [inline transformer_engine/pytorch/attention.py:106]: See https://github.com/NVIDIA/TransformerEngine/pull/1226/files#r1795511669

File: transformer_engine/pytorch/attention.py (hunk 4 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py:125]: I think the difference between pre-2.1 and post-2.1 is the causal/causal_bottom_right mask. Before 2.1, FA only supports top-left causal, and after 2.1, only bottom-right causal. I wonder if there are still users using that.

https://github.com/cyanguwa/TransformerEngine/blob/6fd94c74334800d0488a89d4cf836d96edbf0f1b/transformer_engine/pytorch/attention.py#L643

Comment (sbhavani) [reply transformer_engine/pytorch/attention.py:125]: Should we mention the version of FA we've tested with in the docs somewhere? Related issue: https://github.com/NVIDIA/TransformerEngine/issues/1236

Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py:125]: What do you think of [c2b3be8](https://github.com/NVIDIA/TransformerEngine/pull/1226/commits/c2b3be843aa3bd37a380da2e5446dfb5fce69a39)?

Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:125]: I think this test is too strong, the test suite would grow exponentially especially if we checked each specific version of our dependencies

**Comment by cyanguwa** [comment]
Pipeline 19261498

**Review by ksivaman — APPROVED** [review]
LGTM

**Review by sbhavani — APPROVED** [review]
LGTM

**Comment by cyanguwa** [comment]
Pipeline 19350071, and local tests with the same container for `L1_pytorch_FA_versions_test` are fine (~1h20mins).

**Pull request merged**